node/src/common/JXCoreEngine.cpp

```diff
@@ -0,0 +1,456 @@
+    LogTrace("JXCoreEngine::DefineScriptFile(\"%s\", \"...\")", scriptFileName.c_str());
+
+    if (scriptFileName == mainScriptFileName)
+    {
+        throw new std::invalid_argument("Invalid script file name: 'main.js' is a reserved name.");
+    }
+
+    _dispatcher.Dispatch([this, scriptFileName, scriptCode]()
+    {
+        if (!_started)
+        {
+            _initialScriptMap.emplace(scriptFileName, scriptCode);
+        }
+        else
+        {
+            JX_DefineFile(scriptFileName.c_str(), scriptCode.c_str());
+        }
+    });
+}
+
+void JXCoreEngine::Start(std::string workingDirectory, std::function<void(std::exception_ptr ex)> callback)
+{
+    LogTrace("JXCoreEngine::Start(\"%s\")", workingDirectory.c_str());
+
+    if (workingDirectory.length() == 0)
+    {
+        throw std::invalid_argument("A working directory is required.");
+    }
+
+    if (_workingDirectory.length() == 0)
+    {
+        _workingDirectory = workingDirectory;
+    }
+    else if (workingDirectory != _workingDirectory)
+    {
+        // This limitation of JXCore is not represented in the INodeEngine interface (e.g. as a static
+        // initialization method taking the working directory), because other node engines might not
+        // have the same limitation.
+        LogErrorAndThrow("Cannot start multiple JXCore instances with different working directories.");
+    }
+
+    try
+    {
+        std::call_once(_initOnce, [=]()
+        {
+            JX_InitializeOnce(workingDirectory.c_str());
+        });
+    }
+    catch (...)
+    {
+        LogError("Failed to initialize JXCore engine.");
+        throw;
+    }
+
+    _dispatcher.Dispatch([this, callback]()
+    {
+        try
+        {
+            if (_started)
+            {
+                LogErrorAndThrow("JXCore engine is already started.");
+            }
+
+            JX_InitializeNewEngine();
+            JX_DefineMainFile(mainScriptCode);
+
+            JX_DefineExtension("jxlog", JXLogCallback);
+            JX_DefineExtension("jxcall", JXCallCallback);
+            JX_DefineExtension("jxresult", JXResultCallback);
+            JX_DefineExtension("jxerror", JXErrorCallback);
+
+            for (const std::pair<std::string, std::string>& scriptEntry : _initialScriptMap)
+            {
+                JX_DefineFile(scriptEntry.first.c_str(), scriptEntry.second.c_str());
+            }
+
+            JX_StartEngine();
+
+            for (const std::pair<std::string, std::function<void(std::string)>> callFromScriptEntry : _initialCallFromScriptMap)
+            {
+                this->RegisterCallFromScriptInternal(callFromScriptEntry.first, callFromScriptEntry.second);
+            }
+
+            _callScriptFunction = new JXValue();
+            JX_New(reinterpret_cast<JXValue*>(_callScriptFunction));
+            JX_Evaluate(callScriptFunctionCode, nullptr, reinterpret_cast<JXValue*>(_callScriptFunction));
+
+            _started = true;
+        }
+        catch (...)
```

Comment:
Good to know. Maybe I'm being overly paranoid, but there is some STL code in there that could throw. Generally I'm trying not to let any exceptions escape on the dispatch thread.

---
In reply to: [71087028](https://github.com/openT2T/client/pull/1#discussion_r71087028) [](ancestors = 71087028)